File: src/data/types/performScanResult.ts
import type { Planet } from "./planet.ts";

// Necessary completion: the Scanner GDD amendment names
// `performScan(playerId, dockedPlanetId): { newlyDiscovered: Planet[] }`
// but a bare object return can't also represent this function's several
// documented rejection cases (ship not docked at the given planet, docked
// planet not yet discovered, no scanner owned) as a normal business
// outcome the caller must handle -- same CraftResult/PurchaseResult/
// HireResult/PurchaseShipResult discriminated-union precedent used
// everywhere else in this codebase.
export interface ScanSucceeded {
  scanned: true;
  newlyDiscovered: Planet[];
}

export interface ScanRejected {
  scanned: false;
  reason: string;
}

export type PerformScanResult = ScanSucceeded | ScanRejected;

File: unity/ProfitableCore/Schema/PerformScanResult.cs
namespace Profitable.Core.Schema;

// Ports src/data/types/performScanResult.ts's PerformScanResult =
// ScanSucceeded | ScanRejected discriminated union.
public abstract class PerformScanResult
{
    public abstract bool Scanned { get; }
}

public sealed class ScanSucceeded : PerformScanResult
{
    public override bool Scanned => true;
    public List<Planet> NewlyDiscovered { get; init; } = new();
}

public sealed class ScanRejected : PerformScanResult
{
    public override bool Scanned => false;
    public string Reason { get; init; } = string.Empty;
}
